src/word_search_generator/generate.py

```diff
@@ -18,112 +33,84 @@ def retry(func):
 
     def wrapper(*args, **kwargs):
         attempt = 0
-        while attempt < config.max_fit_tries:
+        while attempt < max_fit_tries:
             try:
                 return func(*args, **kwargs)
             except Exception:
                 attempt += 1
-        return (kwargs["puzzle"], kwargs["key"])
+        return
 
     return wrapper
 
 
-def out_of_bounds(width: int, height: int, position: tuple[int, int]) -> bool:
-    """Check to make sure `position` is still on the current puzzle.
-
-    Args:
-        width (int): Puzzle board width.
-        height (int): Puzzle board height.
-        position (tuple[int, int]): Position to check.
-
-    Returns:
-        bool: If `position` is on the board.
-    """
+def out_of_bounds(puzzle: WordSearch, position: tuple[int, int]) -> bool:
+    """Validate `position` is within the current puzzle bounds."""
+    width = height = puzzle.size
     row, col = position
     if row < 0 or col < 0 or row > height - 1 or col > width - 1:
         return True
     return False
 
 
-def calc_puzzle_size(words: set[str], level: int, size: Optional[int] = None) -> int:
-    """Calculate the puzzle grid size.
-
-    Based on longest word, word count, and level.
-
-    Args:
-        words (set[str]): Words to be placed in the puzzle.
-        level (int): Puzzle difficulty level.
-        size (Optional[int], optional): Puzzle size.
-
-    Returns:
-        int: Puzzle grid size.
-    """
+def calc_puzzle_size(words: Wordlist, level: Sized, size: Optional[int] = None) -> int:
+    """Calculate the puzzle grid size."""
+    all_words = list(word.text for word in words)
+    longest_word_length = len(max(all_words, key=len))
+    shortest_word_length = len(min(all_words, key=len))
     if not size:
-        longest = max(10, len(max(words, key=len)))
-        # calculate multipler for larger word lists so that most have room to fit
-        multiplier = len(words) / 15 if len(words) > 15 else 1
-        size = int(longest + level * 2 * multiplier)
+        longest = max(10, longest_word_length)
+        # calculate multiplier for larger word lists so that most have room to fit
+        multiplier = len(all_words) / 15 if len(all_words) > 15 else 1
+        # level lengths in config.py are nice multiples of 2
+        l_size = log2(len(level)) if level else 1  # protect against log(0) in tests
+        size = round(longest + l_size * 2 * multiplier)
+    else:
+        if size < shortest_word_length:
```

Comment:
@duck57, what are your thoughts on this approach? For example, if the user specifies a puzzle of size 5 but provides a word list with the shortest word being 6 characters long, obviously none of the words will fit into the puzzle. So instead of just throwing an exception, I adjust the puzzle to fit the smallest word and then output a notification to stderr about the change.